packages/dataviews/src/dataviews-layouts/utils/useInfiniteScrollData.ts

```diff
@@ -0,0 +1,277 @@
+	useEffect( () => {
+		if (
+			( currentPage === 1 && ! allLoadedRecords.length ) ||
+			! view.infiniteScrollEnabled
+		) {
+			// First page - replace all data and initialize range
+			const startPosition = 1;
+			const records = shownData.map( ( record, index ) => {
+				const position = view.infiniteScrollEnabled
+					? startPosition + index
+					: undefined;
+				if ( position !== undefined ) {
+					positionMapRef.current.set( getItemId( record ), position );
+				}
+				return {
+					...record,
+					position,
+				};
+			} );
+			setAllLoadedRecords( records );
+
+			if ( records.length > 0 ) {
+				loadedRangeRef.current = {
+					min: Math.min( ...records.map( ( r ) => r.id ) ),
+					max: Math.max( ...records.map( ( r ) => r.id ) ),
+				};
+			}
+		} else {
+			// Subsequent pages - load more data with placeholders
+			setAllLoadedRecords( ( prev ) => {
+				const existingIds = new Set(
+					prev
+						.filter( ( item ): item is Item => item !== null )
+						.map( getItemId )
+				);
+				// Calculate start position based on the highest position already tracked
+				let nextPosition =
+					positionMapRef.current.size > 0
+						? Math.max( ...positionMapRef.current.values() ) + 1
+						: 1;
+
+				const newRecords = shownData
+					.filter(
+						( record ) => ! existingIds.has( getItemId( record ) )
+					)
+					.map( ( record ) => {
+						const itemId = getItemId( record );
+						let position: number | undefined;
+
+						if ( view.infiniteScrollEnabled ) {
+							// Check if this record already has a position
+							const existingPosition =
+								positionMapRef.current.get( itemId );
+							if ( existingPosition !== undefined ) {
+								position = existingPosition;
+							} else {
+								// Assign new position and increment for next record
+								position = nextPosition;
+								positionMapRef.current.set( itemId, position );
+								nextPosition++;
+							}
+						}
+
+						return {
+							...record,
+							position,
+						};
+					} );
+
+				if ( newRecords.length === 0 ) {
+					return prev;
+				}
+
+				// Update the loaded range
+				const allRecords =
+					scrollDirection === 'up'
+						? [ ...newRecords, ...prev ]
+						: [ ...prev, ...newRecords ];
+
+				const allIds = allRecords
+					.filter( ( r ): r is Item => r !== null )
+					.map( ( r ) => r.id );
+				const newMin = Math.min( ...allIds );
+				const newMax = Math.max( ...allIds );
+
+				loadedRangeRef.current = {
+					min: newMin,
+					max: newMax,
+				};
+
+				// Create array with placeholders to maintain positions
+				const result: ( Item | null )[] = [];
+				for ( let id = newMin; id <= newMax; id++ ) {
+					const record = allRecords.find(
+						( r ) => r !== null && r.id === id
+					);
+					result.push( record || null );
+				}
+
+				// Filter to keep only records that should remain visible
+				// Keep items within a certain range of visible entries
+				if ( visibleEntries.length > 0 ) {
```

Comment:
Follow up: could this be a helper function on it's own? `filterByVisibility` for example. I admit, it's hard for me to follow so wondering if extraction + unit test would assist comprehension. Could be just me though 😅